modules/gapi/src/compiler/passes/transformations.cpp

```diff
@@ -0,0 +1,147 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2019 Intel Corporation
+
+#include "precomp.hpp"
+
+#include <ade/util/zip_range.hpp>
+#include <ade/graph.hpp>
+
+#include "api/gcomputation_priv.hpp"
+
+#include "compiler/gmodel.hpp"
+#include "compiler/gmodelbuilder.hpp"
+#include "compiler/passes/passes.hpp"
+#include "compiler/passes/pattern_matching.hpp"
+
+#include <sstream>
+
+namespace cv { namespace gimpl { namespace passes {
+namespace
+{
+// Generates ADE graph from expression-based computation
+std::unique_ptr<ade::Graph> makeGraph(const cv::GComputation& c)
+{
+    std::unique_ptr<ade::Graph> pG(new ade::Graph);
+    ade::Graph& g = *pG;
+
+    cv::gimpl::GModel::Graph gm(g);
+    cv::gimpl::GModel::init(gm);
+    cv::gimpl::GModelBuilder builder(g);
+    auto proto_slots = builder.put(c.priv().m_ins, c.priv().m_outs);
+
+    // Store Computation's protocol in metadata
+    cv::gimpl::Protocol p;
+    std::tie(p.inputs, p.outputs, p.in_nhs, p.out_nhs) = proto_slots;
+    gm.metadata().set(p);
+
+    return pG;
+}
+
+bool tryToSubstitute(ade::Graph& main,
+                     const std::unique_ptr<ade::Graph>& patternG,
+                     const cv::GComputation& substitute)
+{
+    GModel::Graph gm(main);
+
+    // Note: if there are multiple matches, p must be applied several times (outside-scope work)
+    // 1. fina pattern in graph
+    auto match1 = findMatches(*patternG, gm);
+    if (!match1.ok()) {
+        return false;
+    }
+
+    // 2. build substitute graph inside the main graph
+    cv::gimpl::GModelBuilder builder(main);
+    const auto& proto_slots = builder.put(substitute.priv().m_ins, substitute.priv().m_outs);
+    Protocol p;
+    std::tie(p.inputs, p.outputs, p.in_nhs, p.out_nhs) = proto_slots;
```

Comment:
Why not just add Protocol to metadata and reduce number of arguments in matchPatternToSubstitute?

Reply:
Because we build substitute __inside__ the main graph here. If we change its metadata we'll lose original protocol (we can remember it for a while, though) of the original big graph (which can be different, hence pattern is found in the middle). So not to mess with the metadata too much, it's seemed easier to just pass protocol as a  __distinct__ argument

Reply:
Good